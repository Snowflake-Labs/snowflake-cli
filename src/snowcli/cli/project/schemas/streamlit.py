from __future__ import annotations

from snowcli.cli.project.schemas.relaxed_map import FilePath, RelaxedMap
from strictyaml import (
    Optional,
    Seq,
    Str,
)

streamlit_schema = RelaxedMap(
    {
        "name": Str(),
        "stage": Str(),
        "query_warehouse": Str(),
        Optional("main_file", default="streamlit_app.py"): FilePath(),
        Optional("env_file"): FilePath(),
        Optional("pages_dir"): FilePath(),
        Optional("additional_source_files"): Seq(FilePath()),
    }
)
