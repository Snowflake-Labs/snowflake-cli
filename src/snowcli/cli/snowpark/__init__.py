import typer

from snowcli.cli.snowpark.function import app as function_app
from snowcli.cli.snowpark.package import app as package_app
from snowcli.cli.snowpark.procedure import app as procedure_app

app = typer.Typer(
    name="snowpark",
    context_settings={"help_option_names": ["-h", "--help"]},
    help="Manage functions, procedures and Snowpark objects",
)

app.add_typer(function_app)
app.add_typer(package_app)
app.add_typer(procedure_app)
