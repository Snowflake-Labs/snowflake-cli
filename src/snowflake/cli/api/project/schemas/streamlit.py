from __future__ import annotations

from snowflake.cli.api.project.schemas.relaxed_map import FilePath, RelaxedMap
from snowflake.cli.api.project.util import IDENTIFIER
from strictyaml import (
    Optional,
    Regex,
    Seq,
    Str,
)

streamlit_schema = RelaxedMap(
    {
        "name": Str(),
        Optional("database", default=None): Regex(IDENTIFIER),
        Optional("schema", default=None): Regex(IDENTIFIER),
        Optional("stage", default="streamlit"): Str(),
        "query_warehouse": Str(),
        Optional("main_file", default="streamlit_app.py"): FilePath(),
        Optional("env_file"): FilePath(),
        Optional("pages_dir"): FilePath(),
        Optional("additional_source_files"): Seq(FilePath()),
    }
)
