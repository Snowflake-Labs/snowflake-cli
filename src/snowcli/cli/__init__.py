from __future__ import annotations

import logging
import pkgutil
import sys
import typer
from collections.abc import Container
from pathlib import Path

from snowcli import __about__
from snowcli.cli import (
    connection,
    render,
    sql,
    streamlit,
)
from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS
from snowcli.cli.common.snow_cli_global_context import (
    snow_cli_global_context_manager,
    SnowCliGlobalContext,
)
from snowcli.cli.main.snow_cli_main_typer import SnowCliMainTyper
from snowcli.cli import loggers
from snowcli.cli.snowpark import app as snowpark_app
from snowcli.config import AppConfig
from snowcli.output.formats import OutputFormat
from snowcli.connection_config import ConnectionConfigs

app = SnowCliMainTyper()
log = logging.getLogger(__name__)


def _version_callback(value: bool):
    if value:
        typer.echo(f"SnowCLI Version: {__about__.VERSION}")
        raise typer.Exit()


def setup_global_context(debug: bool):
    def modifications(context: SnowCliGlobalContext) -> SnowCliGlobalContext:
        context.enable_tracebacks = debug
        return context

    snow_cli_global_context_manager.update_global_context(modifications)


@app.command()
def login(
    snowsql_config_path: Path = typer.Option(
        "~/.snowsql/config",
        "--config",
        "-c",
        prompt="Path to Snowsql config",
        help="snowsql config file",
    ),
    connection_name: str = typer.Option(
        ...,
        "--connection",
        "-C",
        prompt="Connection name (for entry in snowsql config)",
        help="connection name from snowsql config file",
    ),
):
    """
    Select a Snowflake connection to use with SnowCLI.
    """
    if not snowsql_config_path.expanduser().exists():
        log.error(f"Path to snowsql config does not exist: {snowsql_config_path}")
        raise typer.Abort()

    connection_configs = ConnectionConfigs(snowsql_config_path)
    if not connection_configs.connection_exists(connection_name):
        log.error(
            f"Connection not found in {snowsql_config_path}: {connection_name}. "
            "You can add with `snow connection add`."
        )
        raise typer.Abort()

    cfg = AppConfig()
    cfg.config["snowsql_config_path"] = str(snowsql_config_path.expanduser())
    cfg.config["snowsql_connection_name"] = connection_name
    cfg.save()
    log.info(f"Using connection name {connection_name} in {snowsql_config_path}")
    log.info(f"Wrote {cfg.path}")


@app.command()
def configure(
    environment: str = typer.Option(
        "dev",
        "-e",
        "--environment",
        help="Name of environment (e.g. dev, prod, staging)",
    ),
    database: str = typer.Option(
        ...,
        "--database",
        prompt="Snowflake database",
        help="Snowflake database",
    ),
    schema: str = typer.Option(
        ...,
        "--schema",
        prompt="Snowflake schema",
        help="Snowflake schema",
    ),
    role: str = typer.Option(
        ...,
        "--role",
        prompt="Snowflake role",
        help="Snowflake role",
    ),
    warehouse: str = typer.Option(
        ...,
        "--warehouse",
        prompt="Snowflake warehouse",
        help="Snowflake warehouse",
    ),
):
    """
    Configure an environment to use with your Snowflake connection.
    """
    log.info(f"Configuring environment {environment}...")
    cfg = AppConfig()
    if environment in cfg.config and not typer.confirm(
        "Environment {environment} already exists. Overwrite?",
    ):
        log.error("Cancelling...")
        raise typer.Abort()

    cfg.config[environment] = {}
    cfg.config[environment]["database"] = database
    cfg.config[environment]["schema"] = schema
    cfg.config[environment]["role"] = role
    cfg.config[environment]["warehouse"] = warehouse
    cfg.save()
    log.info(f"Wrote environment {environment} to {cfg.path}")


@app.callback()
def default(
    version: bool = typer.Option(
        None,
        "--version",
        help="Prints version of the snowcli",
        callback=_version_callback,
        is_eager=True,
    ),
    output_format: OutputFormat = typer.Option(
        OutputFormat.TABLE.value,
        "--format",
        help="Specifies output format",
        case_sensitive=False,
        is_eager=True,
    ),
    configuration_file: Path = typer.Option(
        None,
        "--config-file",
        help="Specifies snowcli configuration file that should be used",
        exists=True,
        dir_okay=False,
        is_eager=True,
    ),
    verbose: bool = typer.Option(
        None,
        "--verbose",
        "-v",
        help="Print logs from level info and higher",
    ),
    debug: bool = typer.Option(
        None,
        "--debug",
        help="Print logs from level debug and higher, logs contains additional information",
    ),
) -> None:
    """
    SnowCLI - A CLI for Snowflake
    """
    loggers.create_loggers(verbose, debug)
    setup_global_context(debug=debug)


MODULE_IGNORE_SET = frozenset(("main", "procedure_coverage"))


def register_cli_typers(ignore_container: Container[str] = MODULE_IGNORE_SET) -> None:
    for _, name, _ in pkgutil.walk_packages(__path__):
        if name not in ignore_container:
            cli_app = __import__(f"{__name__}.{name}", fromlist=["_trash"])
            try:
                app.add_typer(cli_app.app, name=name)
            except AttributeError:
                # Ignore modules that don't define app global
                pass


register_cli_typers()

app.command("sql")(sql.execute_sql)
app.add_typer(snowpark_app)


if __name__ == "__main__":
    app()

if getattr(sys, "frozen", False):
    app(sys.argv[1:])
