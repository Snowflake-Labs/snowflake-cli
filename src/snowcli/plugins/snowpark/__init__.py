from snowcli.plugins.snowpark.commands import app
from snowcli.plugins.snowpark.package.commands import app as package_app
from snowcli.plugins.snowpark.procedure_coverage.commands import (
    app as procedure_coverage_app,
)

app.add_typer(package_app)
app.add_typer(procedure_coverage_app)
