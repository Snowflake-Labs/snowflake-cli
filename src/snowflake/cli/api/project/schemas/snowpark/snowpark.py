from __future__ import annotations

from typing import List

from pydantic import Field
from snowflake.cli.api.project.schemas.snowpark.callable import (
    FunctionSchema,
    ProcedureSchema,
)
from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel


class Snowpark(UpdatableModel):
    project_name: str = Field(title="Project identifier")
    stage_name: str = Field(title="Stage in which projectâ€™s artifacts will be stored")
    src: str = Field(title="Folder where your code should be located")
    functions: List[FunctionSchema] | None = Field(
        title="List of functions defined in the project", default=[]
    )
    procedures: List[ProcedureSchema] | None = Field(
        title="List of procedures defined in the project", default=[]
    )
