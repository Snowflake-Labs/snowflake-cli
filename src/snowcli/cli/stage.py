from __future__ import annotations

from pathlib import Path

import typer

from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS, ConnectionOption
from snowcli.snow_connector import SqlExecutionMixin
from snowcli.output.printing import print_db_cursor

app = typer.Typer(
    name="stage",
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    help="Manage stages",
)

StageNameOption = typer.Argument(..., help="Stage name.")


class StageManager(SqlExecutionMixin):
    @staticmethod
    def get_standard_stage_name(name: str) -> str:
        # Handle embedded stages
        if name.startswith("snow://"):
            return name

        return f"@{name}"

    def list(self, stage_name: str):
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(f"ls {stage_name}")

    def get(self, stage_name: str, dest_path: str):
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(f"get {stage_name} file://{dest_path}/")

    def put(self, local_path: str, stage_name: str, parallel: int, overwrite: bool):
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(
            f"put file://{local_path} {stage_name} "
            f"auto_compress=false parallel={parallel} overwrite={overwrite}"
        )

    def remove(self, stage_name: str, path: str):
        stage_name = self.get_standard_stage_name(stage_name)
        return self._execute_query(f"remove {stage_name}{path}")

    def show(self):
        return self._execute_query("show stages")

    def create(self, stage_name: str):
        return self._execute_query(f"create stage if not exists {stage_name}")

    def drop(self, stage_name: str):
        return self._execute_query(f"drop stage {stage_name}")


@app.command("list")
def stage_list(
    connection_name: str = ConnectionOption,
    stage_name=typer.Argument(None, help="Name of stage"),
):
    """
    List stage contents or shows available stages if stage name not provided.
    """
    manager = StageManager.from_connection(connection_name=connection_name)

    if stage_name:
        results = manager.list(stage_name=stage_name)
    else:
        results = manager.show()
    print_db_cursor(results)


@app.command("get")
def stage_get(
    connection_name: str = ConnectionOption,
    stage_name: str = StageNameOption,
    path: Path = typer.Argument(
        Path.cwd(),
        exists=False,
        file_okay=True,
        dir_okay=True,
        writable=True,
        resolve_path=True,
        help="Directory location to store downloaded files",
    ),
):
    """
    Download all files from a stage to a local directory.
    """
    results = StageManager.from_connection(connection_name=connection_name).get(
        stage_name=stage_name, dest_path=path
    )
    print_db_cursor(results)


@app.command("put")
def stage_put(
    connection_name: str = ConnectionOption,
    path: Path = typer.Argument(
        ...,
        exists=True,
        file_okay=True,
        dir_okay=True,
        writable=True,
        resolve_path=True,
        help="File or directory to upload to stage",
    ),
    name: str = StageNameOption,
    overwrite: bool = typer.Option(
        False,
        help="Overwrite existing files in stage",
    ),
    parallel: int = typer.Option(
        4,
        help="Number of parallel threads to use for upload",
    ),
):
    """
    Upload files to a stage from a local client
    """
    manager = StageManager.from_connection(connection_name=connection_name)
    local_path = str(path) + "/*" if path.is_dir() else str(path)

    results = manager.put(
        local_path=local_path, stage_name=name, overwrite=overwrite, parallel=parallel
    )
    print(results)
    print_db_cursor(results)


@app.command("create")
def stage_create(
    connection_name: str = ConnectionOption,
    name: str = StageNameOption,
):
    """
    Create stage if not exists.
    """
    results = StageManager.from_connection(connection_name=connection_name).create(
        stage_name=name
    )
    print_db_cursor(results)


@app.command("drop")
def stage_drop(
    connection_name: str = ConnectionOption,
    name: str = StageNameOption,
):
    """
    Drop stage
    """
    results = StageManager.from_connection(connection_name=connection_name).drop(
        stage_name=name
    )
    print_db_cursor(results)


@app.command("remove")
def stage_remove(
    connection_name: str = ConnectionOption,
    stage_name: str = StageNameOption,
    file_name: str = typer.Argument(..., help="File name"),
):
    """
    Remove file from stage
    """

    results = StageManager.from_connection(connection_name=connection_name).remove(
        stage_name=stage_name, path=file_name
    )
    print_db_cursor(results)
