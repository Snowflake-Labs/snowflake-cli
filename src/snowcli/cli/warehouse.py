from __future__ import annotations

from functools import wraps

import typer
from snowflake.connector.cursor import SnowflakeCursor

from snowcli.cli.common.flags import DEFAULT_CONTEXT_SETTINGS, ConnectionOption
from snowcli.snow_connector import SqlExecutionMixin
from snowcli.output.printing import print_db_cursor

app = typer.Typer(
    name="warehouse",
    context_settings=DEFAULT_CONTEXT_SETTINGS,
    help="Manage warehouses",
)


class WarehouseManager(SqlExecutionMixin):
    def show(self) -> SnowflakeCursor:
        return self._execute_query("show warehouses")


def with_output(func):
    @wraps(func)
    def wrapper(*args, **kwargs):
        result = func(*args, **kwargs)
        print_db_cursor(result)

    return wrapper


@app.command("status")
@with_output
def warehouse_status(connection_name: str = ConnectionOption):
    """
    Show the status of each warehouse in the configured environment.
    """
    results = WarehouseManager.from_connection(connection_name=connection_name).show()
    return results
