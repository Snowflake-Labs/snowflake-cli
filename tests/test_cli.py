from typer.testing import CliRunner

runner = CliRunner()


def test_global(snapshot):
    from snowcli.cli import app

    result = runner.invoke(app, ["-h"])

    assert result.exit_code == 0

    assert result.output == snapshot


def test_streamlit(snapshot):
    from snowcli.cli.streamlit import app

    result = runner.invoke(app, ["-h"])

    assert result.exit_code == 0

    assert result.output == snapshot


def test_stage(snapshot):
    from snowcli.cli.stage import app

    result = runner.invoke(app, ["-h"])

    assert result.exit_code == 0

    assert result.output == snapshot


def test_snowpark(snapshot):
    from snowcli.cli.snowpark import app

    result = runner.invoke(app, ["-h"])

    assert result.exit_code == 0

    assert result.output == snapshot


def test_warehouse(snapshot):
    from snowcli.cli.warehouse import app

    result = runner.invoke(app, ["-h"])

    assert result.exit_code == 0

    assert result.output == snapshot
