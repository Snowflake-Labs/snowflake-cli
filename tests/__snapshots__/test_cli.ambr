# serializer version: 1
# name: test_global
  '''
                                                                                  
   Usage: default [OPTIONS] COMMAND [ARGS]...                                     
                                                                                  
   SnowCLI - A CLI for Snowflake                                                  
                                                                                  
  ╭─ Options ────────────────────────────────────────────────────────────────────╮
  │ --version                                            Prints version of the   │
  │                                                      snowcli                 │
  │ --format                      [TABLE|JSON]           Specifies output format │
  │                                                      [default: TABLE]        │
  │ --config-file                 FILE                   Specifies snowcli       │
  │                                                      configuration file that │
  │                                                      should be used          │
  │                                                      [default: None]         │
  │ --verbose             -v                             Print logs from level   │
  │                                                      info and higher         │
  │ --debug                                              Print logs from level   │
  │                                                      debug and higher, logs  │
  │                                                      contains additional     │
  │                                                      information             │
  │ --install-completion          [bash|zsh|fish|powers  Install completion for  │
  │                               hell|pwsh]             the specified shell.    │
  │                                                      [default: None]         │
  │ --show-completion             [bash|zsh|fish|powers  Show completion for the │
  │                               hell|pwsh]             specified shell, to     │
  │                                                      copy it or customize    │
  │                                                      the installation.       │
  │                                                      [default: None]         │
  │ --help                -h                             Show this message and   │
  │                                                      exit.                   │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  ╭─ Commands ───────────────────────────────────────────────────────────────────╮
  │ configure    Configure an environment to use with your Snowflake connection. │
  │ connection   Manage connection to Snowflake                                  │
  │ login        Select a Snowflake connection to use with SnowCLI.              │
  │ snowpark     Manage functions, procedures and Snowpark objects               │
  │ sql          Executes Snowflake query.                                       │
  │ stage        Manage stages                                                   │
  │ streamlit    Manage Streamlit in Snowflake                                   │
  │ warehouse    Manage warehouses                                               │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  
  '''
# ---
# name: test_snowpark
  '''
                                                                                  
   Usage: snowpark [OPTIONS] COMMAND [ARGS]...                                    
                                                                                  
   Manage functions, procedures and Snowpark objects                              
                                                                                  
  ╭─ Options ────────────────────────────────────────────────────────────────────╮
  │ --install-completion          [bash|zsh|fish|powers  Install completion for  │
  │                               hell|pwsh]             the specified shell.    │
  │                                                      [default: None]         │
  │ --show-completion             [bash|zsh|fish|powers  Show completion for the │
  │                               hell|pwsh]             specified shell, to     │
  │                                                      copy it or customize    │
  │                                                      the installation.       │
  │                                                      [default: None]         │
  │ --help                -h                             Show this message and   │
  │                                                      exit.                   │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  ╭─ Commands ───────────────────────────────────────────────────────────────────╮
  │ function        Manage user defined functions                                │
  │ package         Manage custom Python packages for Snowpark                   │
  │ procedure       Manage stored procedures                                     │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  
  '''
# ---
# name: test_stage
  '''
                                                                                  
   Usage: stage [OPTIONS] COMMAND [ARGS]...                                       
                                                                                  
   Manage stages                                                                  
                                                                                  
  ╭─ Options ────────────────────────────────────────────────────────────────────╮
  │ --install-completion          [bash|zsh|fish|powers  Install completion for  │
  │                               hell|pwsh]             the specified shell.    │
  │                                                      [default: None]         │
  │ --show-completion             [bash|zsh|fish|powers  Show completion for the │
  │                               hell|pwsh]             specified shell, to     │
  │                                                      copy it or customize    │
  │                                                      the installation.       │
  │                                                      [default: None]         │
  │ --help                -h                             Show this message and   │
  │                                                      exit.                   │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  ╭─ Commands ───────────────────────────────────────────────────────────────────╮
  │ create      Create stage if not exists                                       │
  │ drop        Drop stage                                                       │
  │ get         Download files from a stage to a local client                    │
  │ list        List stage contents                                              │
  │ put         Upload files to a stage from a local client                      │
  │ remove      Remove file from stage                                           │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  
  '''
# ---
# name: test_streamlit
  '''
                                                                                  
   Usage: root [OPTIONS] COMMAND [ARGS]...                                        
                                                                                  
   Manage Streamlit in Snowflake                                                  
                                                                                  
  ╭─ Options ────────────────────────────────────────────────────────────────────╮
  │ --install-completion          [bash|zsh|fish|powers  Install completion for  │
  │                               hell|pwsh]             the specified shell.    │
  │                                                      [default: None]         │
  │ --show-completion             [bash|zsh|fish|powers  Show completion for the │
  │                               hell|pwsh]             specified shell, to     │
  │                                                      copy it or customize    │
  │                                                      the installation.       │
  │                                                      [default: None]         │
  │ --help                -h                             Show this message and   │
  │                                                      exit.                   │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  ╭─ Commands ───────────────────────────────────────────────────────────────────╮
  │ create           Create a streamlit app named NAME.                          │
  │ deploy           Deploy streamlit with NAME.                                 │
  │ describe         Describe a streamlit app.                                   │
  │ drop             Create a streamlit app named NAME.                          │
  │ list             List streamlit apps.                                        │
  │ share            Create a streamlit app named NAME.                          │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  
  '''
# ---
# name: test_warehouse
  '''
                                                                                  
   Usage: status [OPTIONS]                                                        
                                                                                  
   Show the status of each warehouse in the configured environment.               
                                                                                  
  ╭─ Options ────────────────────────────────────────────────────────────────────╮
  │ --environment                 TEXT                   Environment name        │
  │                                                      [default: dev]          │
  │ --install-completion          [bash|zsh|fish|powers  Install completion for  │
  │                               hell|pwsh]             the specified shell.    │
  │                                                      [default: None]         │
  │ --show-completion             [bash|zsh|fish|powers  Show completion for the │
  │                               hell|pwsh]             specified shell, to     │
  │                                                      copy it or customize    │
  │                                                      the installation.       │
  │                                                      [default: None]         │
  │ --help                -h                             Show this message and   │
  │                                                      exit.                   │
  ╰──────────────────────────────────────────────────────────────────────────────╯
  
  
  '''
# ---
